tempesta_fw/http_sticky.c

```diff
@@ -0,0 +1,740 @@
+	tfw_http_msg_add_data(&state, msg, buf, len);
+	tfw_http_msg_add_data(&state, msg, S_CRLFCRLF, SLEN(S_CRLFCRLF));
+
+bh_unlock_sock(client->sock); /* TEMPORARY TO SEE IT WORKING */
+	tfw_connection_send_cli(conn, (TfwMsg *)resp);
+	tfw_http_msg_free((TfwHttpMsg *)resp);
+
+	return 0;
+}
+
+#define	S_502_FIXLEN							\
+	SLEN(S_502)							\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_DATE) + SLEN(S_V_DATE)				\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_CONTENT_LENGTH) + SLEN(S_V_CONTENT_LENGTH)		\
+	+ SLEN(S_CRLFCRLF)
+/*
+ * Send an HTTP 502 response to the client. It tells the client that
+ * Tempesta is unable to forward the request to the designated server.
+ */
+int
+tfw_http_send_502(TfwHttpMsg *hm)
+{
+	u_char buf[256];
+	TfwMsg *msg;
+	TfwHttpResp *resp;
+	tfw_mastate_t state = { 0 };
+	size_t len, data_len = S_502_FIXLEN;
+	TfwConnection *conn = hm->conn;
+
+	if ((resp = (TfwHttpResp *) tfw_http_msg_alloc(Conn_Srv)) == NULL) {
+		return -1;
+	}
+	msg = (TfwMsg *)resp;
+
+	if (tfw_http_msg_setup(&state, msg, data_len) != 0) {
+		tfw_http_msg_free((TfwHttpMsg *)resp);
+		return -1;
+	}
+
+	tfw_http_msg_add_data(&state, msg, S_502, SLEN(S_502));
+	tfw_http_msg_add_data(&state, msg, S_CRLF, SLEN(S_CRLF));
+
+	tfw_http_msg_add_data(&state, msg, S_F_DATE, SLEN(S_F_DATE));
+	len = tfw_http_prep_date(buf);
+	tfw_http_msg_add_data(&state, msg, buf, len);
+	tfw_http_msg_add_data(&state, msg, S_CRLF, SLEN(S_CRLF));
+
+	tfw_http_msg_add_data(&state, msg, S_F_CONTENT_LENGTH,
+					   SLEN(S_F_CONTENT_LENGTH));
+	tfw_http_msg_add_data(&state, msg, "0", 1);
+	tfw_http_msg_add_data(&state, msg, S_CRLFCRLF, SLEN(S_CRLFCRLF));
+
+	tfw_connection_send_cli(conn, (TfwMsg *)resp);
+	tfw_http_msg_free((TfwHttpMsg *)resp);
+
+	return 0;
+}
+
+static void
+tfw_http_strim(tfw_str_t *s)
+{
+	u_char *sptr, *eptr;
+
+	if (!s->len)
+		return;
+
+	eptr = s->ptr + s->len - 1;
+	while (eptr >= s->ptr && isspace(*eptr))
+		eptr--;
+	s->len = eptr - s->ptr + 1;
+
+	sptr = skip_spaces(s->ptr);
+	s->len -= sptr - s->ptr;
+	s->ptr = sptr;
+}
+
+static TfwStr *
+tfw_http_field(TfwHttpMsg *hm, const char *field, size_t len)
+{
+	int i;
+	TfwStr *hdr;
+
+	for (i = 0; i < hm->h_tbl->size; i++) {
+		hdr = &hm->h_tbl->tbl[i].field;
+		if (tfw_str_eq_cstr(hdr, field, len,
+				    TFW_STR_EQ_PREFIX | TFW_STR_EQ_CASEI)) {
+			break;
+		}
+	}
+	if (i < hm->h_tbl->size) {
+		/* DIRTY HACK TO COMPENSATE FOR PARSER BUG */
+		const tfw_str_t s_cookie = tfw_string("Cookie:");
+		if (!strncmp(field, s_cookie.ptr, s_cookie.len))
+			hdr->len += 3;
+		return hdr;
+	}
+	return NULL;
+}
+
+static int
+tfw_http_field_val(TfwHttpMsg *hm, const tfw_str_t *field, tfw_str_t *value)
+{
+	char *buf;
+	size_t len;
+	TfwStr *hdr;
+
+	if ((hdr = tfw_http_field(hm, field->ptr, field->len)) == NULL) {
+		return 0;
+	}
+	len = tfw_str_len(hdr) + 1;
+	if ((buf = tfw_pool_alloc(hm->pool, len)) == NULL) {
+		return -ENOMEM;
+	}
+	len = tfw_str_to_cstr(hdr, buf, len);
```

Comment:
I surely agree. The major reason for that was that `TfwStr{}` has no notion of an offset. There's no readily available way to get a substring of a '`TfwStr{}` string, because you can't point at the middle of a chunk in the scope of a `TfwStr{}`. Also, there's not current support for getting an offset of a substring match when doing a substring search, which might also be just because of the same issue.

There are two typical examples that I've encountered here. One is that I need to get the value of a header field: everything minus the actual `"<Header Field Name>: "`. The other is to find a substring, then get another substring that follows (as in <cookie name>=<cookie value>. Provided that we get the position of a matched substring (which we can't right now), that position can occur in the middle of a chunk.

Perhaps we could introduce a notion of an offset into `TfwStr{}`. That would take some extra space in the structure, but it would let us do things described above. Alternatively, we can split a chunk on the fly at an offset. That would require us to extend an array of chunks that make up an `TfwStr{}`, copy all chunks after the offset, and insert a new chunk in the middle that has data right up to the offset.

So, would adding an offset to `TfwStr{}` be a viable solution? Tha would require some work implementing and updating the rest of the codebase, but then we would be able to avoid copying nearly at all times, I believe.